### DISCUSSION TITLE
EC Balancing - All shards on one volumes for a while

### DISCUSSION THREAD

**Discussion by danfoster** [seaweedfs/seaweedfs#5740] [General]
Hi,

I'm trying out the EC sharding and want to clarify my understanding:

I had a replicated volume across 2 volume servers (mode `001`). This test env has 4 volume servers in total.  I then ran `ec.encode` which split it in to the 10+4 shards. But all shared ended up on on a single volume server.

I'm sure that `ec.rebalance` would rebalance the shards. But I'm concerned about the interim time between starting `ec.encode` and `ec.rebalance` finishing. Since the `ec.encode` process deletes the replicas, there is a period where I have all data for that volume on a single disk.

Is this the expected behaviour? If so, are there any recommended strategies for mitigating the risk in this period?

**Comment by chrislusf** [comment]
> But all shared ended up on on a single volume server.

Not the case. see https://github.com/seaweedfs/seaweedfs/blob/master/weed/shell/command_ec_encode.go#L41

```
	This command will:
	1. freeze one volume
	2. apply erasure coding to the volume
	3. move the encoded shards to multiple volume servers
```

**Reply by danfoster** [reply]
I think I see what's happening:

```
markVolumeReadonly 151 on 10.43.0.251:8081 ...
markVolumeReadonly 151 on 10.43.0.251:8083 ...
generateEcShards  151 on 10.43.0.251:8081 ...
parallelCopyEcShardsFromSource 151 10.43.0.251:8081
allocate 151.[0 1 2 3 4 5 6 7 8 9 10 11 12 13] 10.43.0.251:8081 => 10.43.0.251:8083
copy 151.[0 1 2 3 4 5 6 7 8 9 10 11 12 13] 10.43.0.251:8081 => 10.43.0.251:8083
mount 151.[0 1 2 3 4 5 6 7 8 9 10 11 12 13] on 10.43.0.251:8083
I0703 17:35:46.210369 command_ec_common.go:98 10.43.0.251:8081 ec volume 151 deletes shards [0 1 2 3 4 5 6 7 8 9 10 11 12 13]
unmount 151.[0 1 2 3 4 5 6 7 8 9 10 11 12 13] from 10.43.0.251:8081
delete 151.[0 1 2 3 4 5 6 7 8 9 10 11 12 13] from 10.43.0.251:8081
delete volume 151 from 10.43.0.251:8081
delete volume 151 from 10.43.0.251:8083
```

All but one of my volumeServers had no free slots. Which, if I read this function correctly means that all shards get allocated to the same VolumeServer: https://github.com/seaweedfs/seaweedfs/blob/3ebeae0c0ba94659bb39e055ae0278db5bb825f1/weed/shell/command_ec_encode.go#L240

Would it be a good idea to make  the EC process abort if the EC shards are distributed to <4 volumeServers?
